[WIP] iTunes Connect Support

KrauseFx opened the pull request:
This pull request will add iTunes Connect support to `spaceship` :rocket: :boom: 

iTunes Connect Features
- [ ] Apps
  - [x] Create a new app
  - [x] Create new version of an app
  - [x] Update app details
  - [x] Download existing screenshots
  - [x] Submit app for review
  - [x] Add, edit and remove testers
  - [x] Receive App Status (Review, etc.)
  - [x] Receive Rejection Information
  - [x] Manage TestFlight Builds
  - [x] Put builds into Beta Testing
  - [ ] File uploads (especially for the App Icon for both iOS and Apple Watch)

Todos:
- [x] Wrap the model classes into their own modules (`Portal`, `Tunes`)
- [ ] Implement a `reload` method to easily fetch the latest data for a given object (in particular for `Spaceship::Tunes::Application` (@snatchev)
- [x] Implement dot syntax for mapping of results (@snatchev)
- [x] Login error message when accessing `Spaceship::Application.all` without logging in
- [x] Support overriding of setters (for the app's categories to prepend `MZGenre.` before the category name (@snatchev)
- [ ] Tests fail when the order is randomised :disappointed: (@snatchev)
- [ ] Better output when using `bundle console`, to not show garbage

**Later**
- [ ] Add new languages (we do that using the transporter right now)

KrauseFx committed:
Refactored spaceship to support soon support iTC

snatchev commented:
@KrauseFx I am adding this checklist to keep track of the progress as we add support for different iTunes connect features. I am not entirely sure which features you wanted to `spaceship` so feel free to edit this comment.

iTunes Connect Features

- [ ] Apps
  - [ ] Create a new ap
  - [ ] Update app details
  - [ ] Upload screenshots
  - [ ] Submit Xcode build
  - [ ] Submit for review
  - [ ] Add testers

- [ ] Analytics
  - [ ] dump overview metrics within time range 
  - [ ] dump sales metrics
  - [ ] dump usage metrics

KrauseFx committed:
Moved Dev Portal tests into separate folder

KrauseFx committed:
Moved Dev Portal fixtues into Dev Portal folder

KrauseFx committed:
Added iTunes Connect Authentification & Tests

KrauseFx committed:
Fixed fixtures method name conflicts

KrauseFx committed:
Added tests for login to iTunesConnect

KrauseFx committed:
Fixed all tests

KrauseFx committed:
Added inital version of app parsing

KrauseFx committed:
Added fetching and parsing of applications and versions

Added overwriting of setters

KrauseFx committed:
Added application tests

KrauseFx committed:
Added support for multi lang content in version

KrauseFx committed:
Added tests for all the app version and localised things

KrauseFx committed:
Wrapped all the modules into their own modules

while still support legacy code

KrauseFx committed:
Made localized attributes read-only

KrauseFx committed:
Added parsing of more app details and fixing apple's bugs

KrauseFx committed:
Added documentation and tests

KrauseFx committed:
Added parsing of error sections and raising exception

Added tests for that

KrauseFx committed:
Improved error handling of iTC responses

KrauseFx committed:
Added tests for successfull and invalid update metadata requests

KrauseFx committed:
Started with create_version! method

KrauseFx committed:
app.edit_version now returns nil if there is no edit_version

Added tests for that

KrauseFx committed:
raw_data is now always stored automatically

KrauseFx committed:
Implemented create_version and added tests

KrauseFx committed:
Added new find method to application and added tests

KrauseFx committed:
Fixed storing of raw_data in Portal objects

KrauseFx committed:
Added create_application! method and tests

snatchev committed:
introduce DataHash that can lookup nested properties in a Hash.

snatchev committed:
fix exception message

KrauseFx committed:
Merge pull request #59 from fastlane/feature/itc-dot-notation-mapping

introduce DataHash that can lookup nested properties in a Hash.

KrauseFx committed:
Fixed enterprise tests

KrauseFx committed:
Added proper parsing of app status

KrauseFx committed:
Added tests for app icon URL and watch app URL

KrauseFx committed:
Added fetching and parsing of screenshots with tests

KrauseFx committed:
Made screenshots read-only

KrauseFx committed:
Version bump

KrauseFx committed:
Added new language converter class with language mapping

KrauseFx committed:
Screenshot objects now also contain their language

KrauseFx committed:
Added `latest_version` method to application class

KrauseFx committed:
Added string extension for language conversion

KrauseFx committed:
Added resolution center API and tests

KrauseFx commented:
@snatchev just in case you find some time, we'd need a `reload` method to re-download the application from the server. Something like
```
application.reload!
```
which will fetch the latest information from the server. 

We'll need it for
```ruby
      # Create a new version of your app
      def create_version!(version_number)
        if edit_version
          raise "Cannot create a new version for this app as there already is an `edit_version` available"
        end

        client.create_version!(apple_id, version_number)

        # Since we have stored the outdated raw_data, we need to refresh this object
        # otherwise `edit_version` will return nil
        # TODO: implemented -reload method
      end
```

KrauseFx committed:
Moved client_spec

KrauseFx committed:
Fixed .to_json for base class

KrauseFx committed:
Added more tests

KrauseFx committed:
Improved error handling for iTC response data

KrauseFx committed:
Improved error handling

KrauseFx committed:
Added support for release notes with tests

KrauseFx committed:
Added verification if language exists with test

KrauseFx committed:
Added nice inspector for multi language items to better support irb

KrauseFx committed:
Started with creating new languages

mathcarignani committed:
Added suport to wildcard bundle

mathcarignani committed:
Added create method with wildcard test

KrauseFx committed:
Merge pull request #60 from mathcarignani/feature/itc

Itunes Connect Support - Wildcard bundle

KrauseFx committed:
Added initial support for build trains, builds and added tests

KrauseFx committed:
Merge branch 'feature/itc' of https://github.com/fastlane/spaceship into feature/itc

KrauseFx committed:
Removed unneeded reference to BuildTrain

KrauseFx committed:
Fixed inspect of language item when value is nil

KrauseFx committed:
Merge branch 'master' into feature/itc

# Conflicts:
#	lib/spaceship/provisioning_profile.rb
#	lib/spaceship/version.rb

mathcarignani committed:
Added support for external testers

mathcarignani committed:
Working on app testers

mathcarignani committed:
Fixed hash attributes and some code refactoring

KrauseFx committed:
Added parsing and testing of the version number as string

mathcarignani committed:
Added tester subclasses and add/remove methods in application

mathcarignani committed:
Documented tester code

KrauseFx committed:
Disabled adding devices for InHouse profiles

KrauseFx committed:
Added circle.yml

KrauseFx committed:
Added circle.yml

KrauseFx committed:
Added url method to app_version and application

KrauseFx committed:
Added keys and languages method to the LanguageItem

KrauseFx committed:
Fixed unit tests

mathcarignani committed:
Changed tester subclasses logic

KrauseFx committed:
Added parsing and setting of app review notes with tests

snatchev committed:
re-write how the attr_mapping works:

Before: we would define mapped get/set methods on object initialization. This would overwrite any methods of the same name.

After: we extend the class using Module#include.
This allows us to redefine the methods after they have been created by `attr_mapping`. To call the original method, call `super` in the redefinition as if it we a subclass.

snatchev committed:
add test case for extending mapped attributes

mathcarignani committed:
Working on internal testers methods

snatchev committed:
fix a few broken tests where `app` was undefined.

snatchev committed:
fill in review field mapping to make the tests pass

mathcarignani committed:
Added add_all_testers! method to Appplication

mathcarignani committed:
Added missing attributes to tester

KrauseFx committed:
Merge branch 'master' into feature/itc

# Conflicts:
#	circle.yml
#	lib/spaceship/provisioning_profile.rb
#	lib/spaceship/version.rb
#	spec/portal/enterprise_spec.rb

KrauseFx committed:
Added nice error message when using spaceship methods and not being logged in

KrauseFx committed:
Added build train parsing and processing

Support for processing builds
Switched build trains to use the version number as key

KrauseFx committed:
Added more tests for languageItem

KrauseFx committed:
Added putting build trains into beta testing mode

Added tests

KrauseFx committed:
Fixed failing test

snatchev committed:
add a test to make sure our setters are updating the raw_data hash.

KrauseFx committed:
Moved team related code into portal_client

KrauseFx committed:
Merge branch 'feature/itc' of https://github.com/fastlane/spaceship into feature/itc

KrauseFx committed:
Added automatic pre-fixing of category name if necessary

Added tests

KrauseFx committed:
Changed ruby version for circle

KrauseFx commented:
@tobiasstrebitzer Maybe you could take over submitting an app for review? This will probably involve the steps about encryption and so on. Also, you can then also migrate this part in `deliver` if you find the time :+1:

tobiasstrebitzer committed:
implementation of app submission for review & rspec tests

tobiasstrebitzer commented:
@KrauseFx, yes, we're struggling with that anyways at the moment.

I've started on it, and the basics are done. Some scenarios require submission of documents, which is not yet implemented.

Should i create a separate PR, or push to this one?

KrauseFx commented:
@tobiasstrebitzer please submit a pull request to this branch. Don't worry about the document upload, in the beginning we want to cover the standard use cases, which are also supported by `deliver` right now.

tobiasstrebitzer committed:
improved error handling; fixed app submission (start) version data error; more rspec tests

tobiasstrebitzer committed:
Fix app_submission factory: only serve required params for "contentRights" section

KrauseFx committed:
Merge pull request #65 from MagLoft/feature/itc_app_submission

Feature: iTunes Connect - Submit for Review

tobiasstrebitzer committed:
add waitingForReview status mapping to AppStatus.get_from_string method

KrauseFx committed:
Merge pull request #66 from MagLoft/feature/itc_app_status_fix

BugFix: add waitingForReview status mapping to AppStatus.get_from_string method

mathcarignani committed:
Merged with branch feature/itc

KrauseFx committed:
Merge pull request #61 from mathcarignani/feature/itc

[WIP] ITC - External testers

KrauseFx committed:
Added tests for testers

KrauseFx commented:
I'll merge-in the current version to do the initial release for `produce` :+1: Thanks everyone for your help!

KrauseFx commented:
@snatchev @tobiasstrebitzer @mathcarignani please subscribe to https://github.com/fastlane/spaceship/issues/68 instead :+1:
